// @microsoft/loader-set-webpack-public-path/lib/codeGenerator.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
 * See LICENSE in the project root for license information.
 */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var os_1 = require("os");
var SetWebpackPublicPathLoader_1 = require("./SetWebpackPublicPathLoader");
var uglify = require("uglify-js");
function getSetPublicPathCode(options, emitWarning) {
    if (!options.webpackPublicPathVariable) {
        throw '"webpackPublicPathVariable" option must be defined.';
    }
    var lines = [];
    if (options.regexName) {
        // Double-escape backslashes to make them show up as single backslashes in regexps.
        var escapedRegex = options.regexName.replace(/\\/, '\\\\');
        lines = [
            "var scripts = document.getElementsByTagName('script');"
        ];
        var regexInitializationSnippet = "new RegExp('" + escapeSingleQuotes(escapedRegex) + "', 'i')";
        var regexVarName = options.regexVariable;
        if (options.regexVariable) {
            lines.push.apply(lines, [
                "var regex = (typeof " + regexVarName + " !== 'undefined') ? " + regexVarName + " : " + regexInitializationSnippet + ";"
            ]);
        }
        else {
            lines.push.apply(lines, [
                "var regex = " + regexInitializationSnippet + ";"
            ]);
        }
        lines.push.apply(lines, [
            'var found = false;',
            '',
            'if (scripts && scripts.length) {',
            '  for (var i = 0; i < scripts.length; i++) {',
            '    if (!scripts[i]) continue;',
            "    var path = scripts[i].getAttribute('src');",
            '    if (path && path.match(regex)) {',
            "      " + options.webpackPublicPathVariable + " = path.substring(0, path.lastIndexOf('/') + 1);",
            '      found = true;',
            '      break;',
            '    }',
            '  }',
            '}',
            '',
            'if (!found) {',
            "  for (var global in " + SetWebpackPublicPathLoader_1.SetWebpackPublicPathLoader.registryVarName + ") {",
            '    if (global && global.match(regex)) {',
            "      " + options.webpackPublicPathVariable + " = global.substring(0, global.lastIndexOf('/') + 1);",
            '      break;',
            '    }',
            '  }',
            '}'
        ]);
    }
    else {
        if (options.publicPath) {
            lines = [
                "var publicPath = '" + appendSlashAndEscapeSingleQuotes(options.publicPath) + "';"
            ];
        }
        else if (options.systemJs) {
            lines = [
                "var publicPath = window.System ? window.System.baseURL || '' : '';",
                "if (publicPath !== '' && publicPath.substr(-1) !== '/') publicPath += '/';"
            ];
        }
        else {
            emitWarning("Neither 'publicPath' nor 'systemJs' is defined, so the public path will not be modified");
            return '';
        }
        if (options.urlPrefix && options.urlPrefix !== '') {
            lines.push('', "publicPath += '" + appendSlashAndEscapeSingleQuotes(options.urlPrefix) + "';");
        }
        lines.push('', options.webpackPublicPathVariable + " = publicPath;");
    }
    return joinLines(lines, options.linePrefix);
}
exports.getSetPublicPathCode = getSetPublicPathCode;
function getGlobalRegisterCode(debug) {
    if (debug === void 0) { debug = false; }
    var lines = [
        '(function(){',
        "if (!" + SetWebpackPublicPathLoader_1.SetWebpackPublicPathLoader.registryVarName + ") " + SetWebpackPublicPathLoader_1.SetWebpackPublicPathLoader.registryVarName + "={};",
        "var scripts = document.getElementsByTagName('script');",
        'if (scripts && scripts.length) {',
        '  for (var i = 0; i < scripts.length; i++) {',
        '    if (!scripts[i]) continue;',
        "    var path = scripts[i].getAttribute('src');",
        "    if (path) " + SetWebpackPublicPathLoader_1.SetWebpackPublicPathLoader.registryVarName + "[path]=true;",
        '  }',
        '}',
        '})();'
    ];
    var joinedScript = joinLines(lines);
    if (debug) {
        return "" + os_1.EOL + joinedScript;
    }
    else {
        var uglified = uglify.parse(joinedScript);
        uglified.figure_out_scope();
        var compressor = uglify.Compressor({
            dead_code: true
        });
        var compressed = uglified.transform(compressor);
        compressed.figure_out_scope();
        compressed.compute_char_frequency();
        compressed.mangle_names();
        return "" + os_1.EOL + compressed.print_to_string();
    }
}
exports.getGlobalRegisterCode = getGlobalRegisterCode;
function joinLines(lines, linePrefix) {
    return lines.map(function (line) {
        if (!!line) {
            return "" + (linePrefix || '') + line;
        }
        else {
            return line;
        }
    }).join(os_1.EOL).replace(new RegExp("" + os_1.EOL + os_1.EOL + "+", 'g'), "" + os_1.EOL + os_1.EOL);
}
function escapeSingleQuotes(str) {
    if (str) {
        return str.replace('\'', '\\\'');
    }
    else {
        return undefined;
    }
}
function appendSlashAndEscapeSingleQuotes(str) {
    if (str && str.substr(-1) !== '/') {
        str = str + '/';
    }
    return escapeSingleQuotes(str);
}
